proc/sort/sort.go

```diff
@@ -55,112 +56,140 @@ func (p *Proc) Done() {
 	p.parent.Done()
 }
 
-func (p *Proc) sortLoop() {
+func (p *Proc) run() {
 	defer close(p.resultCh)
-	firstRunRecs, eof, err := p.recordsForOneRun()
-	if err != nil || len(firstRunRecs) == 0 {
-		p.sendResult(nil, err)
-		return
-	}
-	p.setCompareFn(&firstRunRecs[0])
-	if eof {
-		// Just one run so do an in-memory sort.
-		expr.SortStable(firstRunRecs, p.compareFn)
-		//XXX bug: we need upstream ectx. See #3367
-		array := zbuf.NewArray(firstRunRecs)
-		p.appendWarnings(array)
-		p.sendResult(array, nil)
-		return
-	}
-	// Multiple runs so do an external merge sort.
-	runManager, err := p.createRuns(firstRunRecs)
-	if err != nil {
-		p.sendResult(nil, err)
-		return
-	}
-	defer runManager.Cleanup()
-	puller := zbuf.NewPuller(runManager, 100)
-	for p.pctx.Err() == nil {
-		// Reading from runManager merges the runs.
-		b, err := puller.Pull()
-		p.sendResult(b, err)
-		if b == nil || err != nil {
-			if err == nil {
-				warnings := zbuf.NewArray(nil)
-				p.appendWarnings(warnings)
-				p.sendResult(warnings, nil)
-			}
-			return
-		}
-	}
-}
-
-func (p *Proc) sendResult(b zbuf.Batch, err error) {
-	select {
-	case p.resultCh <- proc.Result{Batch: b, Err: err}:
-	case <-p.pctx.Done():
-	}
-}
-
-func (p *Proc) recordsForOneRun() ([]zed.Value, bool, error) {
+	var spiller *spill.MergeSort
+	var eof bool
 	var nbytes int
 	var out []zed.Value
 	for {
 		batch, err := p.parent.Pull()
 		if err != nil {
-			return nil, false, err
+			p.sendResult(nil, err)
+			return
```

Comment:
If `spiller` is non-nil, this will return without calling `spiller.Cleanup`. But rather than insert a call here, I think you should simplify the protocol: Remove all the existing calls (including the one in `sendSpills`). Add one call immediately before `spiller = nil`. And add a deferred call immediately after `var spiller *spill.MergeSort`:
```go
	 defer func() {
	 	 if spiller != nil {
	 	 	 spiller.Cleanup()
		}
	}()
```

Reply:
good idea!

Reply:
The cleanup in sendSpills is needed on success because otherwise the tmp files will never be deleted.  Maybe Cleanup() should be called Close()?

Reply:
That's why I suggested adding one call immediately before `spiller = nil`. Keeping the call in `sendSpills` as you have makes the protocol less clear—readers have to examine both methods to see that `spiller.Cleanup` is called appropriately.

Reply:
Ok, fixed